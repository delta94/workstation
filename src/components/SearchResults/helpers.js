import walkme from '@walkme/sdk';
import uniqWith from 'lodash/uniqWith';
import isEqual from 'lodash/isEqual';

function findInNode(node, searchTerm) {
  const description = node.description ?? '';
  const url = node.properties.url ?? '';
  const keywords = node.keywords?.join(' ') ?? '';

  return [node.title, description, url, keywords].some((string) =>
    string.toLowerCase().includes(searchTerm.toLowerCase())
  );
}

function flattenUITree(uiTree) {
  const flattenedContent = [];

  uiTree
    .map((tab) => tab.childNodes)
    .flat(1)
    .forEach((node) => {
      node.type === 'category' ? flattenedContent.push(...node.childNodes) : flattenedContent.push(node);
    });

  return uniqWith(flattenedContent, isEqual);
}

export function findInUiTree(searchTerm, uiTree) {
  return (
    flattenUITree(uiTree)
      .filter((node) => ![walkme.content.TypeNames.Survey].includes(node.type))
      // get items that include the search term in their attributes
      .filter((node) => findInNode(node, searchTerm))
  );
}

export async function findInSearchApi(searchTerm, wmSearch) {
  try {
    const searchResults = await wmSearch.search(searchTerm);
    return searchResults.filter((item) => item.type === walkme.content.TypeNames.SearchResult);
  } catch (err) {
    console.error('Could not get search results from the API.', err);
  }
}
