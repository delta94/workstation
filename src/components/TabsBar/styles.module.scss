.tabs {
  position: relative;

  ul.tabs-list {
    width: 100%;
    height: 80px;
    margin: 0;
    padding: 0 30px;
    background-color: var(--secondary100);
    border-top: 1px solid var(--border500);
    border-bottom: 1px solid var(--border500);
    list-style-type: none;
    font-size: 16px;
    display: flex;

    li.tab {
      height: 100%;
      line-height: 80px;
      margin-right: 56px;
      padding-right: 2px;
      cursor: pointer;
      color: var(--secondary500);
      font-family: var(--font-secondary), sans-serif;
      display: inline-flex;
      align-items: center;
      transition: all 250ms ease;
      min-width: 0;

      &:last-child {
        margin-right: 0;
      }

      &.active {
        color: var(--primary);
      }

      &:hover {
        margin-top: -2px;
        color: var(--primary);
      }

      svg {
        margin-right: 11px;
        flex-shrink: 0;
      }

      .tab-title {
        white-space: nowrap;
        overflow: hidden;
        text-overflow: ellipsis;
      }
    }
  }

  .tabs-underline {
    height: 3px;
    background-color: var(--primary);
    width: var(--underline-width);
    position: absolute;
    left: var(--underline-left);
    bottom: -1px;
    transition: all 200ms cubic-bezier(0.68, -0.2, 0.265, 1.2);
  }
}
