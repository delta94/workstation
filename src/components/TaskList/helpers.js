function flattenNodes(content) {
  const flattenedContent = [];

  content.forEach((node) => {
    node.type === 'category' ? flattenedContent.push(...node.childNodes) : flattenedContent.push(node);
  });

  return flattenedContent;
}

export function calculateCompletedPercent(content) {
  const flatTasks = flattenNodes(content);
  const total = flatTasks.length;
  const completed = flatTasks.filter((node) => node.properties.isCompleted).length;
  return Math.round((completed / total) * 100);
}
