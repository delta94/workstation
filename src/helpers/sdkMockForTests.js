const searchResults = [
  {
    title: 'Peace to Prosperity',
    description:
      'Jan 1, 2020 ... CONTENTS. GLOSSARY | 1 ... adjustments to the share of Egyptian, Israeli, or \nPalestinian content required to meet the 35 percent content.',
    type: 'search-result',
    properties: {
      knowledgeBaseType: 'Google',
      url: 'http://pic-upload.ynet.co.il/news/Peace-to-Prosperity-0120.pdf',
    },
  },
  {
    title: 'USDA National Fluoride Database of Selected Beverages and...',
    description:
      'Fluoride content of municipal water in the United States: What percentage is \nfluoridated? J. Food Comp. Anal. 16(5): 621-628. Ophaug RH. 1983-1987.',
    type: 'search-result',
    properties: {
      knowledgeBaseType: 'Google',
      url: 'http://go.ynet.co.il/pic/briut/F02.pdf',
    },
  },
  {
    title: 'Fat and Energy Contents of Expressed Human Breast Milk in...',
    description:
      'Energy contents of the milk were measured with a bomb calorimeter. Results. \nThe groups did not differ in terms of mater- nal height and diet, infant birth weight,\n ...',
    type: 'search-result',
    properties: {
      knowledgeBaseType: 'Google',
      url: 'http://prdupl02.ynet.co.il/ForumFiles/13424549.pdf',
    },
  },
  {
    title: 'אמבלייז רוכשת השליטה ב-Smart Content מרעננה',
    description:
      '16 ינואר 2006 ... Smart Content מפתח טכנולוגיה ל"דחיפת" תוכן מותאם אישית למכשירים ניידים ומחשבים \nאישיים על גבי רשתות אלחוטיות ו-IP. אמבלייז מתכוונת לשלב את טכנולוגיית ...',
    type: 'search-result',
    properties: {
      knowledgeBaseType: 'Google',
      url: 'https://www.ynet.co.il/articles/0,7340,L-3201198,00.html',
    },
  },
  {
    title: 'מפת הבחירות של ישראל 2019 - ynet',
    description:
      'המידע המוצג מבוסס על עיבוד שערך אתר ״מדלן - הכל על בתים ושכונות״ לנתוני ועדת הבחירות \nהמרכזית. נתוני השכונות מקורבים לפי התאמת קלפיות. כל המידע מוצג תחת תנאי השימוש ...',
    type: 'search-result',
    properties: {
      knowledgeBaseType: 'Google',
      url: 'https://z.ynet.co.il/short/content/ElectionMap2019/',
    },
  },
  {
    title: 'X-rays',
    description:
      'hirty years ago, X-rays were the only way to see what was going on inside your \nbody. Now other methods of medical imaging are available, some using different\n ...',
    type: 'search-result',
    properties: {
      knowledgeBaseType: 'Google',
      url: 'http://prdupl02.ynet.co.il/ForumFiles_2/16567358.pdf',
    },
  },
  {
    title: 'Vizual.ai',
    description:
      'Content Publishers, Photo Editors & Online Retailers can now leverage the \npower of the Vizual.ai platform to select the optimal imagery for their customers.',
    type: 'search-result',
    properties: {
      knowledgeBaseType: 'Google',
      url: 'https://yq-api.ynet.co.il/',
    },
  },
  {
    title: 'גוגל: בטא של תוכנית Adsense for Content בעברית',
    description:
      '27 דצמבר 2006 ... השירות של גוגל, המאפשר הצבת מודעות קשורות תוכן, הושק בעברית בבטא, בשלב ראשון רק \nבאתר תפוז.',
    type: 'search-result',
    properties: {
      knowledgeBaseType: 'Google',
      url: 'https://www.ynet.co.il/articles/0,7340,L-3345160,00.html',
    },
  },
  {
    title: '(function() { CmmAppVideoApi.DecorationTypes = { Credit: ...',
    description:
      '_registerContentEvents(content); this.pool.push(content); }; }; this.\n_registerContentEvents = function(content){ var contentEvents = content.\ngetEventList(); ...',
    type: 'search-result',
    properties: {
      knowledgeBaseType: 'Google',
      url: 'https://www.ynet.co.il/Cmn/App/Video/CmmAppVideoApi_Js_Content/0,13943,280716135620,00.html',
    },
  },
  {
    title: 'תחרות בתחומי העיצוב וה-e-Content',
    description:
      '23 יוני 2006 ... Top Talent של EUROPRIX היא תחרות אירופית עבור סטודנטים ואנשי מקצוע צעירים, \nהעוסקים בפרויקטים חדשניים בתחומי העיצוב וה- e-Content (תוכן אלקטרוני), ...',
    type: 'search-result',
    properties: {
      knowledgeBaseType: 'Google',
      url: 'https://www.ynet.co.il/articles/0,7340,L-3266501,00.html',
    },
  },
];

const sdk = {
  wmSearch: {
    userGuid: '0854dff97dca4aa18d85358b00cd9593',
    enduserGuid: 'user@walkme.com',
    serverPath: 'https://playerserver.walkmeqa.com',
    language: null,
    envId: 0,
    saveSearch: false,
    search: jest.fn(() => searchResults),
  },
  wmNotifications: {
    _notificationRecurrences: {
      consts: {
        RECURRENCE_TYPES: {
          ONCE: 'Once',
          DAY: 'Daily',
          WEELK: 'Weekly',
        },
        RECURRENCE_CODES: {
          ONCE: 0,
          DAY: 1,
          WEEK: 2,
        },
      },
    },
    getNotifications() {
      return this.notifications;
    },
  },
  notifications: [
    {
      type: 'notification',
      id: 'aaa',
      title: 'Notification 1',
      description: 'Message 1',
      properties: {
        isPlayed: true,
        isSilent: true,
        date: '2020-06-10T07:01:58.900Z',
      },
    },
    {
      type: 'notification',
      id: 'bbb',
      title: 'Notification 2',
      description: 'Message 2',
      properties: {
        isPlayed: false,
        isSilent: false,
        date: '2020-06-10T07:01:58.900Z',
      },
    },
    {
      type: 'notification',
      id: 'ccc',
      title: 'Notification 3',
      description: 'Message 3',
      properties: {
        isPlayed: false,
        isSilent: false,
        date: '2020-06-10T07:01:58.900Z',
      },
    },
  ],
  uiTreeSDK: [
    {
      type: 'tab',
      id: 762,
      title: 'Assistant Home',
      description: 'Description for tab-762',
      properties: {
        tabType: 'help',
      },
      childNodes: [
        {
          type: 'smart-walkthru',
          id: 280,
          title: 'Content smart-walkthru-280',
          description: 'Description for smart-walkthru-280',
          properties: {},
        },
        {
          type: 'survey',
          id: 777,
          title: 'Content survey-777',
          description: 'Description for survey-777',
          properties: {},
        },
        {
          type: 'shuttle',
          id: 27,
          title: 'Content shuttle-27',
          description: 'Description for shuttle-27',
          properties: {
            url: 'https://27.example.com/index?t=599',
          },
        },
        {
          type: 'video',
          id: 981,
          title: 'Content video-981',
          description: 'Description for video-981',
          properties: {},
        },
        {
          type: 'article',
          id: 723,
          title: 'Content article-723',
          description: 'Description for article-723',
          properties: {},
        },
        {
          type: 'category',
          id: 20,
          title: 'Category',
          description: 'Description for category-20',
          childNodes: [
            {
              type: 'smart-walkthru',
              id: 554,
              title: 'Content smart-walkthru-554',
              description: 'Description for smart-walkthru-554',
              properties: {},
            },
            {
              type: 'survey',
              id: 211,
              title: 'Content survey-211',
              description: 'Description for survey-211',
              properties: {},
            },
            {
              type: 'shuttle',
              id: 70,
              title: 'Content shuttle-70',
              description: 'Description for shuttle-70',
              properties: {
                url: 'https://27.example.com/index?t=599',
              },
            },
            {
              type: 'video',
              id: 896,
              title: 'Content video-896',
              description: 'Description for video-896',
              properties: {},
            },
            {
              type: 'article',
              id: 522,
              title: 'Content article-522',
              description: 'Description for article-522',
              properties: {},
            },
          ],
          properties: {},
        },
      ],
    },
    {
      type: 'tab',
      id: 313,
      title: 'Ongoing Tasks',
      description: 'Description for tab-313',
      properties: {
        tabType: 'tasks',
      },
      childNodes: [
        {
          type: 'task',
          id: 989,
          title: 'Not completed',
          description: 'Description for task-989',
          properties: {
            isCompleted: false,
            isCrossedOff: false,
            isDisabled: false,
          },
        },
        {
          type: 'task',
          id: 892,
          title: 'Not completed',
          description: 'Description for task-892',
          properties: {
            isCompleted: false,
            isCrossedOff: false,
            isDisabled: false,
          },
        },
        {
          type: 'task',
          id: 528,
          title: 'Completed - not crossed off, not disabled',
          description: 'Description for task-528',
          properties: {
            isCompleted: true,
            isCrossedOff: false,
            isDisabled: false,
          },
        },
        {
          type: 'task',
          id: 690,
          title: 'Completed - not crossed off, not disabled',
          description: 'Description for task-690',
          properties: {
            isCompleted: true,
            isCrossedOff: false,
            isDisabled: false,
          },
        },
        {
          type: 'task',
          id: 483,
          title: 'Completed - crossed off, not disabled',
          description: 'Description for task-483',
          properties: {
            isCompleted: true,
            isCrossedOff: true,
            isDisabled: false,
          },
        },
        {
          type: 'task',
          id: 243,
          title: 'Completed - crossed off, not disabled',
          description: 'Description for task-243',
          properties: {
            isCompleted: true,
            isCrossedOff: true,
            isDisabled: false,
          },
        },
        {
          type: 'task',
          id: 79,
          title: 'Completed - crossed off, disabled',
          description: 'Description for task-79',
          properties: {
            isCompleted: true,
            isCrossedOff: true,
            isDisabled: true,
          },
        },
        {
          type: 'task',
          id: 963,
          title: 'Completed - crossed off, disabled',
          description: 'Description for task-963',
          properties: {
            isCompleted: true,
            isCrossedOff: true,
            isDisabled: true,
          },
        },
        {
          type: 'task',
          id: 596,
          title: 'Completed - not crossed off, disabled',
          description: 'Description for task-596',
          properties: {
            isCompleted: true,
            isCrossedOff: false,
            isDisabled: true,
          },
        },
        {
          type: 'task',
          id: 462,
          title: 'Completed - not crossed off, disabled',
          description: 'Description for task-462',
          properties: {
            isCompleted: true,
            isCrossedOff: false,
            isDisabled: true,
          },
        },
        {
          type: 'category',
          id: 140,
          title: 'Day 1',
          description: 'Description for category-140',
          childNodes: [
            {
              type: 'task',
              id: 196,
              title: 'Not completed',
              description: 'Description for task-196',
              properties: {
                isCompleted: false,
                isCrossedOff: false,
                isDisabled: false,
              },
            },
            {
              type: 'task',
              id: 694,
              title: 'Not completed',
              description: 'Description for task-694',
              properties: {
                isCompleted: false,
                isCrossedOff: false,
                isDisabled: false,
              },
            },
            {
              type: 'task',
              id: 646,
              title: 'Completed - not crossed off, not disabled',
              description: 'Description for task-646',
              properties: {
                isCompleted: true,
                isCrossedOff: false,
                isDisabled: false,
              },
            },
            {
              type: 'task',
              id: 194,
              title: 'Completed - not crossed off, not disabled',
              description: 'Description for task-194',
              properties: {
                isCompleted: true,
                isCrossedOff: false,
                isDisabled: false,
              },
            },
            {
              type: 'task',
              id: 660,
              title: 'Completed - crossed off, not disabled',
              description: 'Description for task-660',
              properties: {
                isCompleted: true,
                isCrossedOff: true,
                isDisabled: false,
              },
            },
            {
              type: 'task',
              id: 129,
              title: 'Completed - crossed off, not disabled',
              description: 'Description for task-129',
              properties: {
                isCompleted: true,
                isCrossedOff: true,
                isDisabled: false,
              },
            },
            {
              type: 'task',
              id: 103,
              title: 'Completed - crossed off, disabled',
              description: 'Description for task-103',
              properties: {
                isCompleted: true,
                isCrossedOff: true,
                isDisabled: true,
              },
            },
            {
              type: 'task',
              id: 331,
              title: 'Completed - crossed off, disabled',
              description: 'Description for task-331',
              properties: {
                isCompleted: true,
                isCrossedOff: true,
                isDisabled: true,
              },
            },
            {
              type: 'task',
              id: 748,
              title: 'Completed - not crossed off, disabled',
              description: 'Description for task-748',
              properties: {
                isCompleted: true,
                isCrossedOff: false,
                isDisabled: true,
              },
            },
            {
              type: 'task',
              id: 38,
              title: 'Completed - not crossed off, disabled',
              description: 'Description for task-38',
              properties: {
                isCompleted: true,
                isCrossedOff: false,
                isDisabled: true,
              },
            },
          ],
          properties: {},
        },
      ],
    },
    {
      type: 'tab',
      id: 298,
      title: 'Bookmarks',
      description: 'Description for tab-298',
      childNodes: [
        {
          type: 'shuttle',
          id: 518,
          title: 'Content shuttle-518',
          description: 'Description for shuttle-518',
          properties: {
            url: 'https://27.example.com/index?t=599',
          },
        },
        {
          type: 'shuttle',
          id: 492,
          title: 'Content shuttle-492',
          description: 'Description for shuttle-492',
          properties: {
            url: 'https://27.example.com/index?t=599',
          },
        },
        {
          type: 'shuttle',
          id: 971,
          title: 'Content shuttle-971',
          description: 'Description for shuttle-971',
          properties: {
            url: 'https://27.example.com/index?t=599',
          },
        },
        {
          type: 'shuttle',
          id: 708,
          title: 'Content shuttle-708',
          description: 'Description for shuttle-708',
          properties: {
            url: 'https://27.example.com/index?t=599',
          },
        },
        {
          type: 'shuttle',
          id: 132,
          title: 'Content shuttle-132',
          description: 'Description for shuttle-132',
          properties: {
            url: 'https://27.example.com/index?t=599',
          },
        },
        {
          type: 'shuttle',
          id: 381,
          title: 'Content shuttle-381',
          description: 'Description for shuttle-381',
          properties: {
            url: 'https://27.example.com/index?t=599',
          },
        },
        {
          type: 'shuttle',
          id: 304,
          title: 'Content shuttle-304',
          description: 'Description for shuttle-304',
          properties: {
            url: 'https://27.example.com/index?t=599',
          },
        },
        {
          type: 'shuttle',
          id: 353,
          title: 'Content shuttle-353',
          description: 'Description for shuttle-353',
          properties: {
            url: 'https://27.example.com/index?t=599',
          },
        },
        {
          type: 'shuttle',
          id: 548,
          title: 'Content shuttle-548',
          description: 'Description for shuttle-548',
          properties: {
            url: 'https://27.example.com/index?t=599',
          },
        },
        {
          type: 'shuttle',
          id: 481,
          title: 'Content shuttle-481',
          description: 'Description for shuttle-481',
          properties: {
            url: 'https://27.example.com/index?t=599',
          },
        },
      ],
      properties: {
        tabType: 'help',
      },
    },
  ],
  tabTypes: {
    Help: 'help',
    Tasks: 'tasks',
  },
  languagesSDK: [
    {
      displayName: 'default',
      shortName: '',
    },
    {
      displayName: 'spanish',
      shortName: 'es-es',
    },
  ],
  platform: {
    isWindows: false,
    isMac: false,
    isWeb: true,
    isMock: false,
    platform: 'Web',
    platformTypes: ['Windows', 'Mac', 'Web', 'Mock'],
  },
};

const sdkProvider = {
  state: {
    sdk,
    ui: {
      location: {
        contentType: 'help',
        index: 0,
      },
      previousLocation: {
        contentType: 'help',
        index: 0,
      },
      loading: false,
      error: false,
    },
  },
  dispatch: () => {},
};

export { sdkProvider };
