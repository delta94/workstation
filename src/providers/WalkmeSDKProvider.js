import React, { createContext, useEffect, useState } from 'react';
import walkme from '@walkme/sdk';

export const WalkmeSDKContext = createContext(null);

export default function WalkmeSDKProvider({ children }) {
  const [sdk, setSdk] = useState(null);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(false);

  useEffect(() => {
    (async function init() {
      try {
        await walkme.init();
        const [wmSearch, wmNotifications, uiTreeSDK, languagesSDK] = await Promise.all([
          walkme.apps.getApp('search'),
          walkme.apps.getApp('notifications'),
          walkme.content.getContentUITree(),
          walkme.language.getLanguagesList(),
        ]);
        setSdk({ wmSearch, wmNotifications, uiTreeSDK, languagesSDK });
        setLoading(false);
      } catch (error) {
        setLoading(false);
        setError(true);
      }
    })();
  }, []);

  return <WalkmeSDKContext.Provider value={sdk}>{sdk ? children : 'Loading'}</WalkmeSDKContext.Provider>;
}
