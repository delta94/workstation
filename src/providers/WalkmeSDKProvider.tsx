import React, { createContext, useEffect, useState } from 'react';
import walkme from '@walkme/sdk';

type AppPropTypes = { children: React.ReactNode };
type WalkmeSdkContextTypes = {
  wmSearch: object;
  wmNotifications: object;
  notifications: object;
  uiTreeSDK: object;
  tabTypes: object;
  languagesSDK: object;
};

export const WalkmeSDKContext = createContext<WalkmeSdkContextTypes | null>(null);

export default function WalkmeSDKProvider({ children }: AppPropTypes) {
  const [sdk, setSdk] = useState<WalkmeSdkContextTypes | null>(null);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(false);

  useEffect(() => {
    (async function init() {
      try {
        await walkme.init();
        const [wmSearch, wmNotifications, uiTreeSDK, languagesSDK] = await Promise.all([
          walkme.apps.getApp('search'),
          walkme.apps.getApp('notifications'),
          walkme.content.getContentUITree(),
          walkme.language.getLanguagesList(),
        ]);
        const notifications = await wmNotifications.getNotifications();
        const tabTypes = (walkme.content as any).TabTypes;

        setLoading(false);
        setSdk({
          wmSearch,
          wmNotifications,
          notifications,
          uiTreeSDK,
          tabTypes,
          languagesSDK,
        });
      } catch (error) {
        setLoading(false);
        setError(true);
      }
    })();
  }, []);

  return <WalkmeSDKContext.Provider value={sdk}>{sdk ? children : 'Loading'}</WalkmeSDKContext.Provider>;
}
