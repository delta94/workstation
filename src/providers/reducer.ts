// @ts-ignore
import cloneDeep from 'lodash/cloneDeep';
import { Action, AppState } from './types';

export const UPDATE_SDK = 'updateSdk';
export const UPDATE_SDK_SUCCESS = 'updateSdkSuccess';
export const UPDATE_SDK_ERROR = 'updateSdkError';
export const INIT_LOCATION = 'initLocation';
export const UPDATE_LOCATION_HISTORY = 'updateLocationHistory';
export const UPDATE_LOCATION = 'updateLocation';
export const TOGGLE_LOCATION = 'toggleLocation';
export const HANDLE_BEFORE_CLOSE = 'handleBeforeClose';
export const NOTIFICATIONS_RECEIVED = 'notificationsReceived';
export const LOADING_NOTIFICATIONS = 'loadingNotifications';

export const initialState: AppState = {
  sdk: {
    wmSearch: {},
    wmNotifications: {},
    notifications: [],
    isLoadingNotifications: false,
    uiTreeSDK: {},
    tabTypes: {},
    languagesSDK: {},
    platform: {},
  },
  ui: {
    loading: true,
    error: false,
  },
};

export default function reducer(state: AppState, action: Action): AppState {
  const draft = cloneDeep(state);

  switch (action.type) {
    case UPDATE_SDK:
      return { ...draft, ui: { ...draft.ui, loading: true, error: false } };
    case UPDATE_SDK_SUCCESS:
      return { ...draft, sdk: action.sdk, ui: { ...draft.ui, loading: false, error: false } };
    case UPDATE_SDK_ERROR:
      console.error(action.error);
      return { ...draft, ui: { ...draft.ui, loading: false, error: true } };
    case INIT_LOCATION:
      return { ...draft, ui: { ...draft.ui, location: action.location, previousLocation: action.location } };
    case UPDATE_LOCATION_HISTORY:
      return { ...draft, ui: { ...draft.ui, location: action.location, previousLocation: draft.ui.location } };
    case UPDATE_LOCATION:
      return { ...draft, ui: { ...draft.ui, location: action.location } };
    case TOGGLE_LOCATION:
      return {
        ...draft,
        ui: { ...draft.ui, location: draft.ui.previousLocation, previousLocation: draft.ui.location },
      };
    case HANDLE_BEFORE_CLOSE:
      return { ...draft, ui: { ...draft.ui, loading: true, error: false } };

    case LOADING_NOTIFICATIONS:
      return {
        ...draft,
        sdk: {
          ...draft.sdk,
          isLoadingNotifications: true,
        },
      };
    case NOTIFICATIONS_RECEIVED:
      return {
        ...draft,
        sdk: {
          ...draft.sdk,
          isLoadingNotifications: false,
          notifications: action.notifications ?? [],
        },
      };
    default:
      return state;
  }
}
