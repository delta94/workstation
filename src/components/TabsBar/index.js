import React, { useContext, useState, useRef, useEffect } from 'react';
import cc from 'classcat';

import { WalkmeSDKContext } from '../../providers/WalkmeSDKProvider';

import { ReactComponent as AssistantHome } from './icons/assistant-home.svg';
import { ReactComponent as Bookmarks } from './icons/bookmarks.svg';
import { ReactComponent as OngoingTasks } from './icons/ongoing-tasks.svg';
import classes from './styles.module.scss';

const iconsToTabsDictionary = {
  'assistant-home': AssistantHome,
  bookmarks: Bookmarks,
  'ongoing-tasks': OngoingTasks,
};

export default function TabsBar({ onSelectSection, isActive }) {
  const { uiTreeSDK: tabs } = useContext(WalkmeSDKContext);
  const [activeTab, setActiveTab] = useState(undefined);
  const [underlineSizes, setUnderlineSize] = useState(undefined);
  const tabRefs = useRef([new Array(tabs.length)]);

  function onClickTab(tab, index) {
    setActiveTab(tab);
    onSelectSection({ contentType: tab.properties.tabType, content: tab.childNodes, data: tab });

    const { width, left } = tabRefs.current[index].getBoundingClientRect();
    setUnderlineSize({ width, left });
  }

  useEffect(() => {
    if (!isActive) {
      setUnderlineSize({ width: 0, left: 0 });
    }
  }, [isActive]);

  return (
    <section className={classes.tabs}>
      <ul className={classes['tabs-list']}>
        {tabs.map((tab, index) => {
          const tabIconName = tab.title.toLowerCase().replace(' ', '-');
          const TabIcon = iconsToTabsDictionary[tabIconName];
          const isActive = tab === activeTab;

          return (
            <li
              className={cc([classes.tab, { [classes.active]: isActive }])}
              key={tab.id}
              title={tab.description}
              onClick={() => onClickTab(tab, index)}
              ref={(ref) => {
                tabRefs.current[index] = ref;
              }}
            >
              <TabIcon />
              {tab.title}
            </li>
          );
        })}
      </ul>
      <div
        className={classes['tabs-underline']}
        style={{
          '--underline-width': `${underlineSizes?.width ?? 0}px`,
          '--underline-left': `${underlineSizes?.left ?? 0}px`,
        }}
      />
    </section>
  );
}
