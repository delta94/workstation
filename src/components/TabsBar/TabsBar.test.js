import React from 'react';
import { render } from '@testing-library/react';

import { WalkmeSDKContext } from '../../providers/WalkmeSDKProvider';
import TabsBar from './index';

import { sdkProvider } from '../../helpers/sdkMockForTests';

describe('TabsBar renders correctly', () => {
  test('should render correctly', async () => {
    const { container } = render(
      <WalkmeSDKContext.Provider value={sdkProvider}>
        <TabsBar path={{ index: 0 }} isActive={false} />
      </WalkmeSDKContext.Provider>
    );

    expect(container).toMatchSnapshot();
  });
});
