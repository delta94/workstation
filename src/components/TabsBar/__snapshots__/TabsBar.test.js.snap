// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabsBar renders correctly should render correctly 1`] = `
<div>
  <section
    class="tabs"
    data-testid="tabs-bar"
  >
    <ul
      class="tabs-list"
    >
      <li
        class="tab"
        title="Description for tab-762"
      >
        <svg>
          assistant-home.svg
        </svg>
        <span
          class="tab-title"
        >
          Assistant Home
        </span>
      </li>
      <li
        class="tab"
        title="Description for tab-313"
      >
        <svg>
          ongoing-tasks.svg
        </svg>
        <span
          class="tab-title"
        >
          Ongoing Tasks
        </span>
      </li>
      <li
        class="tab"
        title="Description for tab-298"
      >
        <svg>
          bookmarks.svg
        </svg>
        <span
          class="tab-title"
        >
          Bookmarks
        </span>
      </li>
    </ul>
    <div
      class="tabs-underline"
    />
  </section>
</div>
`;
